Do not report "baseline" on formatting information request

CyrilleB79 opened the pull request:
### Link to issue number:

Fixes #11815
Fixes #11078.

### Summary of the issue:

When fetching formatting information with NVDA+F (once or twice) in a Firefox page, "baseline" is reported.
Only subscript or superscript should be reported, not "baseline" since "baseline" should be considered the default text position.
Also, when "Report superscript and subscript" option is checked, navigating in a Firefox page cause spurious reporting of 'baseline'. This is due to transitions between elements having no formatting information such as button or graphics and baseline text.

### Description of how this pull request fixes the issue:

Updated design after review comments in this PR:
* Ensured that all object generating the textInfo format field provide only standard values for 'text-position' field; superscript, subscript, baseline and unspecified.
* Unspecified is used in the following cases:
  - the API does not provide formatting info (e.g. buttons in Firefox)
  - mixed text-position information, e.g. an Excel cell containing baseline and superscript.

### Alternative solution

The old design, initially implemented in this issue before review comments, was the following:
* No modification was done on the values produced by the various API.
* In `getFormatFieldSpeech`in `speech\__init__.py`, when getting 'text-position' attribute, replace any attribute that do not match 'sub' or 'super' with None.
* This allows not to report "baseline" when requesting formatting information, but also not to report "baseline" when the cursor moves from a web element that do not have any formatting attribute available.

### Testing performed:

For the folling APIs/files, the following test were made:
- Check that NVDA+F is able to report superscript and subscript
- Check that NVDA+F does not report text position information for baseline text or unspecified text position (if applicable)
- Activate "Report superscript and subscript" option and navigate whith the caret to test all possible transitions between the 3 or 4 possible text position (superscript, subscript, baseline and unspecified).

API/files tested:
* `NVDAObjects/IAccessible/__init__.py`: tested with Chrome
* `NVDAObjects/UIA/__init__.py`: tested with Word (UIA)
* `NVDAObjects/window/edit.py`, class `ITextDocumentTextInfo`: Tested with Wordpad.
* `NVDAObjects/window/edit.py`, class `EditTextInfo`: I do not know where I can test it natively, but I have used the suggestion made in https://github.com/nvaccess/nvda/pull/11897#issuecomment-932776162 to test (successfully) in WordPad.
* `NVDAObjects/window/excel.py`: tested with Excel, cell navigation.
  Note: When editing the cell's content, no formatting can be reported (neither before this PR nor with this PR).
* `NVDAObjects/window/winword.py`: tested with Word '16.0.5188.1000' (UIA not checked)
* `appModules/powerpnt.py`: tested with PowerPoint slide writing mode (formatting not reported in slideshow mode, neither before this PR nor after).
* `appModules/soffice.py`: tested with Libre Office Writer 7.1
* `virtualBuffers/MSHTML.py`: Tested with IE, with CSS `vertical-align` property.
  Note: `<sub>` / `<super>` tags do not work, neither with this PR nor before.

API/files not (yet) tested:
* `NVDAObjects/JAB/__init__.py`
In which control/program could I test?

Other tests:
* Firefox: With "Report superscript and subscript" option checked, navigate with the virtual cursor on a GitHub issue page around buttons and graphics. No spurious "baseline" can be heard.
* Excel: While editing cell content and pressing NVDA+F, "No formatting information" is now reported, since it cannot be retrieved. Before this PR, "baseline" was reported whatever the content of the cell.

### Known issues with pull request:

Two APIs have not been tested (cf Test section).
Reviewers: please help me complete the test coverage.

Also reviewer, please check that I haven't forgotten any API.

### Change log entry:

Changes
`'Baseline' is not reported anymore upon text formatting request (NVDA+F). (#11815)``

Bug fixes
`'Baseline' is not frequently spuriously reported anymore whild reading some webpages when report superscript and subscript option is active. (#11078)`

CyrilleB79 committed:
When requesting text formatting information, "baseline" is not reported anymore since it is considered as the default text position.
It is still reported on cursor movement if report superscript and subscript option is activated in document formatting options.

feerrenrut commented on source/speech/__init__.py:
```diff
@@ -2210,9 +2210,13 @@ def getFormatFieldSpeech(  # noqa: C901
 		textPosition=attrs.get("text-position")
+		if textPosition:
+			textPosition = textPosition.lower()
+		if textPosition not in ['super', 'sub']:
+			textPosition = None
```
Please add a comment here about the other values it might be, and why NVDA should not report them.

CyrilleB79 replied:
Well, that's not so clear, there are many cases.
* The following object's textInfo may report sub / super / [no field]
  * JABTextInfo (in NVDAObjects/JAB/__init__.py)
  * EditTextInfo (in sub / super / [no field])
  * ITextDocumentTextInfo (in NVDAObjects/window/edit.py:)
  * TextFrameTextInfo (in appModules/powerpnt.py:)
  * WordDocumentTextInfo (in NVDAObjects/window/winword.py: thanks to NVDAHelper code)
* The following object's textInfo may report sub / super / baseline / [no field]
  * UIATextInfo (in NVDAObjects/UIA/__init__.py:)
  * SymphonyTextInfo (in appModules/soffice.py:)
* IA2TextTextInfo (in NVDAObjects\IAccessible\__init__.py:) seems to be able to produce various values that are directly passed. I do not know this part of code tho know where it initially comes from. But the values I have already gotten are: 'auto', 'middle', '50%'.

And this list may not be exhaustive. I have made it by using:
`git grep -i "text-position"`
`git grep -i "reportSuperscriptsAndSubscripts"`
and some additional tests in IE where I had already encountered some strange values.

Anyway, my approach was to follow what was done previously, except for reporting 'baseline', thus leading to the same result if 'baseline' or nothing or any value different from 'sub' or 'super is provided.

If required, please give me further tips to have an exhausting list of where these value can come from and to know the real origin of the value provided by IA2TextTextInfo.
Thanks.

feerrenrut replied:
Ok perhaps cleaning that up may best as a separate task. It would be nice if each of the providers for this info normalized the values to a some Enum value.

@michaelDCurran , do you have any suggestions for ensuring we have a "known set" of values for `textPosition`?

feerrenrut reviewed (changes requested):
Generally this looks good, thanks @CyrilleB79. Add in that comment and I think this is good to go.

CyrilleB79 commented:
Regarding IA2TextTextInfo, the function `normalizeIA2TextFormatField` seems to perform some sort of normalization on the attribute but superscript / subscript seems not to be considered there.

What I can suggest for this PR is at least:
* remove the capability to produce 'baseline' in UIATextInfo and in SymphonyTextInfo. Indeed, this PR considers in `getFormatFieldSpeech` that the text is baseline if `attrs["text-position"] == 'baseline'` or if the attribute `attrs["text-position"]` does not exists at all.
* log a warning in if attrs["text-position"] is something else than 'sub' or 'super', in order to fix this unexpected value in a future PR

Optionally I may also normalize the text-position attribute of IA2TextTextInfo to 'sub' or 'super' only, depending to your comments on this topic.
On [this page](https://wiki.linuxfoundation.org/accessibility/iaccessible2/textattributes),  I have seen that only sub, super and baseline values are allowed. Since the unexpected values ('auto', 'middle', '50%'.) come from IE, maybe it corresponds to an older spec or maybe IE does not fully support the spec? Or do you have another link regarding this spec with more info? I am not sure to get the information at the correct place.

At last, in future PRs, I may fix some other issues encountered thanks to the warning I will have added in this PR.
I have also seen some other potential issues while digging for this PR (WordPad, Excel cells). I will investigate and open new issues/PR for this if required.

michaelDCurran commented:
To clarify, Internet Explorer does not use IAccessible2. Rather we get the value for text-position directly from the CSS vertical-align property: see https://developer.mozilla.org/en-US/docs/Web/CSS/vertical-align for a list of all possible vertical-align values.
I would suggest for IE that we filter out anything that is not super or sub.
This can be done in virtualBuffers.mshtml.mshtmlTextInfo._normalizeFormatField.
In general though, I would a gree the best approach is for all APIs to expose text-position as either only super, sub, or not at all.

CyrilleB79 committed:
Merge branch 'master' into noBaselineInFormattingInfo2

CyrilleB79 committed:
Support only 'sup' and 'super' for 'textInfo' attribute; issue a warning for other values. No 'text-info' attribute at all is expected for baseline (being default value).

CyrilleB79 committed:
APIs do not expose 'baseline' as 'text-position' attribute anymore (UIATextInfo , SymphonyTextInfo).

CyrilleB79 committed:
MSHTML virtual buffer: filter out any CSS text-position attribute other than 'sub' or 'super'.

CyrilleB79 commented:
I have addressed the majority of the comments.

Two points / questions are still pending:

1. @feerrenrut, you suggested in a line comment:
> It would be nice if each of the providers for this info normalized the values to a some Enum value.
As seen with MSHTML (Internet Explorer) virtual buffer, there are some case where the values are passed directly from the API. The enum will not be used there. Is there a risk to mix-up values from the enum and values produced directly as strings?

2. I hope to have considered all the possible APIs that may produce a 'text-position' attribute in a formatted text to speak. Reviewer, please double check it; I do not have the full picture of all the APIs and what they may generate.
Searching for "text-position" in the code should be quite exhaustive however.
In any case, if there is an unexpected value, it will be signaled by the warning.

feerrenrut commented on source/virtualBuffers/MSHTML.py:
```diff
@@ -44,6 +44,10 @@ def _normalizeFormatField(self, attrs):
 		language=attrs.get('language')
 		if language:
 			attrs['language']=languageHandler.normalizeLanguage(language)
+		textPosition = attrs.get('textPosition')
+		if textPosition and textPosition not in ['super', 'sub']:
+			del attrs['text-position']
+		log.debug(f'{attrs}')
```
Would you consider making text position an Enum, and extracting this to a separate method so it works in the same way as `def _getIsCurrentAttribute`  (line 53)?

Then type checking can ensure that the value that shows up in the speech module is only one of the expected values. Someone wishing to add a new value in the future can more easily follow the enum type through the code.

CyrilleB79 replied:
Do you mean using a new `controlTypes.TextPosition` enum for all the text-position format fields passed from the various APIs to `getFormatFieldSpeech`?
If yes:
* would you recommend to define this enum with only 2 values (sub and super)? This would lead to a code a bit strange mixing up the enum values with None in some cases (baseline or other value or missing value).
* On the opposit, we could have an enum with 3 values (sub, super and baseline). But it does not follow anymore the suggestion made by @michaelDCurran in https://github.com/nvaccess/nvda/pull/11897#issuecomment-813802247.
In any case, the code of all the APIs would need to be rework to normalize the values to the enum's one. It is actually better than today's code since an unexpected value coming from an API is checked closer to the source.

feerrenrut replied:
Yes, I personally think that three explicit values makes more sense. Note, that two strings and `None` is also three known values, they just don't share a consistent type.

For each value there should be a mapping to the speech for that value, the enum `BASE_LINE` value should have an empty string (or maybe `None`) for it's speech, take a look at how `isCurrent` works.

feerrenrut replied:
This is still unresolved @CyrilleB79.

CyrilleB79 replied:
> This is still unresolved @CyrilleB79.

Yes, I need to work on this but I need to take a little time to see and check all the cases.
Putting this PR in draft for now.

CyrilleB79 committed:
Merge branch 'master' into noBaselineInFormattingInfo2

CyrilleB79 committed:
Merge branch 'master' into noBaselineInFormattingInfo2

CyrilleB79 committed:
Merge branch 'master' into noBaselineInFormattingInfo2

CyrilleB79 committed:
Addressed review comments.

lukaszgo1 commented:
Hello @CyrilleB79 I was pretty excited when I saw this fix since #11078 is pretty annoying for me however it seems this PR is not progressing. Is it just a matter of finding time to work on this for you or have you encountered any technical difficulties. Is there anything I can help with to progress this work?

CyrilleB79 commented:
Hello

Sorry for the delay. And thanks to ping me.
This PR requires a little time to be updated to the last code base and for testing all the cases and I have postponed it to handle more straight forward issues.
I will try to dedicate time to it and ask for help if needed (maybe for testing the various applications/interfaces).
There should not be technical issue either.

CyrilleB79 committed:
Merge branch 'master' into noBaselineInFormattingInfo2

CyrilleB79 commented:
Re-testing after the upmerges and fixes pushed on 22nd September 2021:
The tests are done with report superscript and subscript activated:

Work in progress:
1. Firefox does not report spuriously 'baseline' when navigating in this GitHub PR page (as previously described in #11078).

2. Checked the transition between superscript, subscript, baseline (and unspecified if applicable) and checked that NVDA+F does not report baseline anymore:
   * Excel '16.0.14326.20404': transitions between cells: 4 types of cell (4th type = cell with mixed text position)
   * Word '16.0.14326.20404' (thus with UIA now): 3 text position types
   * PowerPoint '16.0.14326.20404': 3 text position types (slide edition ; slideshow does not seem to support such report)
   * WordPad (thus IAccessible): 3 text position types

CyrilleB79 committed:
Fixed return to baseline in Chrome.

CyrilleB79 committed:
Fix for Libre Office.

CyrilleB79 commented:
Hello

I have done the rework to generate standard text-position values at each API level and have tried to test all possible cases.
I have updated the initial description according to the new design and with the many additional manual tests done.

@feerrenrut you may review it again since you had asked to use the Enum. Note that @seanbudd was automatically requested as a reviewer when I have switched back this PR from draft to ready state.

@lukaszgo1 if you want to add your review comments or additional test coverage, you are welcome.

At last, sorry for the delay to address this PR review comments. I hope to have it finalized soon now!

lukaszgo1 commented:
> API/files not (yet) tested:
> 
>     * `NVDAObjects/JAB/__init__.py`

It would be pretty tough to find a example for this - the only known JAB object which exposed formatting info was editor in Open Office before they switched to IA2.

> 
>     * `NVDAObjects/window/edit.py`, class `EditTextInfo`
>       In which control/program could I test?

You can always modify `_get_TextInfo` on `Edit` class to unconditionally return `EditTextInfo` and test in wordpad.

CyrilleB79 commented:
> It would be pretty tough to find a example for this - the only known JAB object which exposed formatting info was editor in Open Office before they switched to IA2.

Do you have the version number, and even a portable version to download?

> You can always modify `_get_TextInfo` on `Edit` class to unconditionally return `EditTextInfo` and test in wordpad.

Thanks for the trick. I have been able to test this part of code successfully and have updated the initial description accordingly.

lukaszgo1 commented:
> Do you have the version number, and even a portable version to download?


 It looks like version 3.3.0 still required JAB these builds can be downloaded from http://archive.apache.org/dist/incubator/ooo/stable/3.3.0/. It looks like portable versions weren't officially available.

CyrilleB79 commented:
> It looks like version 3.3.0 still required JAB these builds can be downloaded from http://archive.apache.org/dist/incubator/ooo/stable/3.3.0/. It looks like portable versions weren't officially available.

Thanks @lukaszgo1 for the version number.
Would you have the possibility to test on an installed version? I.e. do you have a virtual machine setup just to install and test?
Feel free to decline however.

lukaszgo1 commented:
> Would you have the possibility to test on an installed version? I.e. do you have a virtual machine setup just to install and test? Feel free to decline however.



I don't have much time now I'll try to look into this during the weekend.

CyrilleB79 commented:
Many thanks Lukasz for your help.

After thinking again to it, Open Office v3.3 is already ten year old. And I do not know any other application using this interface and featuring text with superscript / subscript.
Maybe an attentive review of `NVDAObjects/JAB/__init__.py` without test would be enough; the code is not complex.
@feerrenrut just let us know if you expect to have this part of code tested or not.

lukaszgo1 commented:
It turns out that while OO 3.3 indeed requires JAB to be accessible the version of Java it bundles is too old to work with the version of Access Bridge which is bundled in NVDA. It should probably be possible to work around this somehow, perhaps by forcing Open Office to use never JDK, but given age of this version I don't think trying to force it to work with  NVDA is good use of anyone's time.

CyrilleB79 commented:
@feerrenrut is this PR still on your radar?
Any chance to have it in 2022.1? If yes, please tag it with the milestone.

Having this review finalized would also allow me to submit another work regarding formatting review, using the same strategy.
Thanks.

CyrilleB79 committed:
Merge branch 'master' into noBaselineInFormattingInfo2

CyrilleB79 commented:
@feerrenrut now that 2021.3.1 is released, is there a chance to have this PR reviewed for 2022.1?
There has just been a request again on a French user mailing list on this topic.

CyrilleB79 commented:
@feerrenrut, sorry to insist on this PR.
Even if you have not (yet) time to re-review it, can you at least give a feedback to know if it will be included in 2022.1 or in another release later?
* If 2022.1 is targetted, I have to remove the API-compatibility code since we are already in 2022.1 dev cycle.
* If a later release is targetted, I have to modify API-compatibility code to test against 2023.1 instead of 2022.1.
I may also have to document API compatibility in the change log according to this decision.

Also Cc @seanbudd and @michaelDCurran if they can say about the targetted release.
Many thanks!

seanbudd commented:
I haven't reviewed the code, I have just noticed that a file had a chmod operation, `source/speech/__init__.py  100644 → 100755`. Can this be removed?

CyrilleB79 committed:
Change file permissions.

CyrilleB79 commented:
Thanks for this remark @seanbudd .

The following commits have changed the permissions on this file: b95743da959148924183effa98ffcc7f6e5f0349 and 57f2791654b3e10d929680c769c9d2ab09b5c2b3. They do not belong to this PR, but come from a merge of master branch.
But strangely, the merge of the commits from master has not modified the permissions in my branch. I do not know why.

Anyway, it should be fixed now.

Regarding the targetted release, do you have any information?

CyrilleB79 committed:
Remove unneeded compa code.

CyrilleB79 committed:
Merge branch 'master' into noBaselineInFormattingInfo2

CyrilleB79 commented:
I have hidden my previous comment (https://github.com/nvaccess/nvda/pull/11897#issuecomment-1026590165) regarding compatibility and have tagged it as off-topic. Indeed, it was irrelevant since there is actually no compatibility issue in this PR.
I have just pushed a commit removing the irrelevant code dealing with compatibility since it was not needed.
Sorry for the confusion.

CyrilleB79 committed:
Restore TextPosition that was erroneously removed from controlTypes.__all__

feerrenrut reviewed (approved):
Thanks @CyrilleB79

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into noBaselineInFormattingInfo2

feerrenrut committed:
update changes file for PR #11897
